example-apps/token_curated_registry/auth/handler.go

```diff
@@ -0,0 +1,320 @@
+		commitMsg := msg.(types.CommitMsg)
+
+		store := ctx.KVStore(ballotKey)
+		key := []byte(commitMsg.Identifier)
+		bz := store.Get(key)
+
+		if bz == nil {
+			return sdk.NewError(2, 108, "Candidate with given identifier does not exist").Result()
+		}
+		candidate := &types.Ballot{}
+		err2 := cdc.UnmarshalBinary(bz, candidate)
+		if err2 != nil {
+			panic(err2)
+		}
+
+		if candidate.EndCommitBlockStamp == 0 || candidate.EndCommitBlockStamp < ctx.BlockHeight() {
+			return sdk.NewError(2, 112, "Candidate not in commit phase").Result()
+		}
+
+		commitStore := ctx.KVStore(commitKey)
+
+		voter := types.Voter{
+			Owner:      commitMsg.Owner,
+			Identifier: commitMsg.Identifier,
+		}
+		voterKey, _ := cdc.MarshalBinary(voter)
+		commitStore.Set(voterKey, commitMsg.Commitment)
+		return sdk.Result{}
+	}
+}
+
+func NewRevealHandler(accountKeeper bank.Keeper, ballotMapper db.BallotMapper) sdk.Handler {
+	return func(ctx sdk.Context, msg sdk.Msg) sdk.Result {
+		revealMsg := msg.(types.RevealMsg)
+		_, _, err := accountKeeper.SubtractCoins(ctx, revealMsg.Owner, []sdk.Coin{revealMsg.Bond})
+		if err != nil {
+			return err.Result()
+		}
+
+		store := ctx.KVStore(ballotMapper.BallotKey)
+		key := []byte(revealMsg.Identifier)
+		bz := store.Get(key)
+
+		if bz == nil {
+			return sdk.NewError(2, 108, "Candidate with given identifier does not exist").Result()
+		}
+		candidate := &types.Ballot{}
+		err2 := ballotMapper.Cdc.UnmarshalBinary(bz, candidate)
+		if err2 != nil {
+			panic(err2)
+		}
+
+		if candidate.EndCommitBlockStamp > ctx.BlockHeight() || candidate.EndRevealBlockStamp < ctx.BlockHeight() {
+			return sdk.NewError(2, 112, "Candidate not in reveal phase").Result()
+		}
+
+		commitStore := ctx.KVStore(ballotMapper.CommitKey)
+		revealStore := ctx.KVStore(ballotMapper.RevealKey)
+
+		voter := types.Voter{
+			Owner:      revealMsg.Owner,
+			Identifier: revealMsg.Identifier,
+		}
+		voterKey, _ := ballotMapper.Cdc.MarshalBinary(voter)
+		if revealStore.Get(voterKey) != nil {
+			return sdk.NewError(2, 128, "Cannot vote more than once").Result()
+		}
+
+		commitment := commitStore.Get(voterKey)
+
+		hasher := sha256.New()
+		vz, _ := ballotMapper.Cdc.MarshalBinary(revealMsg.Vote)
+		hasher.Sum(vz)
+		val := hasher.Sum(revealMsg.Nonce)
+
+		if !reflect.DeepEqual(val, commitment) {
+			return sdk.NewError(2, 106, "Vote does not match commitment").Result()
+		}
+
+		reveal := types.Vote{
+			Choice: revealMsg.Vote,
+			Power:  revealMsg.Bond.Amount,
+		}
+		revealVal, _ := ballotMapper.Cdc.MarshalBinary(reveal)
+
+		commitStore.Delete(voterKey)
+		revealStore.Set(voterKey, revealVal)
+
+		err3 := ballotMapper.VoteBallot(ctx, revealMsg.Owner, revealMsg.Identifier, revealMsg.Vote, revealMsg.Bond.Amount)
+		if err2 != nil {
+			return err3.Result()
+		}
+		return sdk.Result{}
+	}
+}
+
+func NewApplyHandler(accountKeeper bank.Keeper, ballotMapper db.BallotMapper, listingKey sdk.StoreKey, quorum float64, dispPct float64) sdk.Handler {
```

Comment:
Should change ballots to be stored in queue, so that they can be processed in EndBlocker efficiently rather than creating a separate message.